File: Pacman-Game-OOP-master 2/EASendMail/Samples_ASPNetMvc/VB/WebProject1/WebProject1/Services/AsyncSendMailService.vb
Imports System.Threading.Tasks
Imports EASendMail

Public Class AsyncSendMailService
    Private Shared _taskStatus As Dictionary(Of String, AsyncSendMailStatus) = Nothing
    Private Shared _lock As Object = Nothing

    Public Shared Sub Init()
        _lock = New Object()
        _taskStatus = New Dictionary(Of String, AsyncSendMailStatus)()
    End Sub

    Public Shared Sub CreateAsyncTask(ByVal myTask As MailTask)
        SyncLock _lock
            Dim status As AsyncSendMailStatus = New AsyncSendMailStatus()
            status.TaskId = myTask.TaskId

            If Not _taskStatus.ContainsKey(myTask.TaskId) Then
                _taskStatus.Add(myTask.TaskId, status)
                Task.Run(
                    Sub()
                        _sendEmail(status, myTask)
                    End Sub
                    )
            End If
        End SyncLock
    End Sub

    Public Shared Sub PutErrorStatus(ByVal taskId As String, ByVal [error] As String)
        SyncLock _lock
            Dim status As AsyncSendMailStatus = New AsyncSendMailStatus()
            status.TaskId = taskId
            status.HasError = True
            status.Status = [error]
            status.Completed = True

            If Not _taskStatus.ContainsKey(taskId) Then
                _taskStatus.Add(taskId, status)
            End If
        End SyncLock
    End Sub

    Public Shared Function QueryStatus(ByVal taskId As String) As AsyncSendMailStatus
        Dim task As AsyncSendMailStatus = Nothing

        SyncLock _lock

            If _taskStatus.ContainsKey(taskId) Then
                task = _taskStatus(taskId)

                If task.Completed Then
                    _taskStatus.Remove(taskId)
                End If
            End If
        End SyncLock

        Return task
    End Function

    Private Shared Sub _sendEmail(ByVal status As AsyncSendMailStatus, ByVal myTask As MailTask)
        Try
            Dim server = myTask.BuildServer()
            Dim mail = myTask.BuildMail()

            Dim smtp = New SmtpClient()
            ' EASendMail event handler
            AddHandler smtp.OnAuthorized, AddressOf OnAuthorized
            AddHandler smtp.OnConnected, AddressOf OnConnected
            AddHandler smtp.OnSecuring, AddressOf OnSecuring
            AddHandler smtp.OnSendingDataStream, AddressOf OnSendingDataStream

            status.Status = "Connecting server ..."
            smtp.Tag = status

            smtp.SendMail(server, mail)

            status.Status = "Completed"
            status.HasError = False
            status.Progress = 100
        Catch ep As Exception
            status.Status = ep.Message
            status.HasError = True
            status.Progress = 100
        End Try

        status.Completed = True
    End Sub

#Region "EASendMail Event Handler"

    Public Shared Sub OnSecuring(ByVal sender As Object, ByRef cancel As Boolean)
        Dim status = CType((CType(sender, SmtpClient)).Tag, AsyncSendMailStatus)
        status.Status = "Securing ..."
    End Sub

    Public Shared Sub OnAuthorized(ByVal sender As Object, ByRef cancel As Boolean)
        Dim status = CType((CType(sender, SmtpClient)).Tag, AsyncSendMailStatus)
        status.Status = "Authorized"
    End Sub

    Public Shared Sub OnConnected(ByVal sender As Object, ByRef cancel As Boolean)
        Dim status = CType((CType(sender, SmtpClient)).Tag, AsyncSendMailStatus)
        status.Status = "Connected"
    End Sub

    Public Shared Sub OnSendingDataStream(ByVal sender As Object, ByVal sent As Integer, ByVal total As Integer, ByRef cancel As Boolean)
        Dim status = CType((CType(sender, SmtpClient)).Tag, AsyncSendMailStatus)

        If total = 0 Then
            Return
        End If

        Dim progress As Integer = (sent * 100) / total
        status.Progress = progress
        status.Status = String.Format("Sending {0}% ...", progress)
    End Sub

#End Region

End Class

File: Pacman-Game-OOP-master 2/EASendMail/Samples_ASPNetMvc/C#/WebProject1/WebProject1/Services/AsyncSendMailService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using System.Threading.Tasks;
using WebProject1.Models;
using EASendMail;

namespace WebProject1.Services
{
    public class AsyncSendMailService
    {
        private static Dictionary<string, AsyncSendMailStatus> _taskStatus = null;
        private static object _lock = null;

        public static void Init()
        {
            _lock = new object();
            _taskStatus = new Dictionary<string, AsyncSendMailStatus>();
        }

        public static void CreateAsyncTask(MailTask mailTask)
        {
            lock (_lock)
            {
                AsyncSendMailStatus status = new AsyncSendMailStatus();
                status.TaskId = mailTask.TaskId;
                if (!_taskStatus.ContainsKey(mailTask.TaskId))
                {
                    _taskStatus.Add(mailTask.TaskId, status);

                    Task.Run(() =>
                    {
                        _sendEmail(status, mailTask);
                    });
                }
            }
        }

        public static void PutErrorStatus(string taskId, string error)
        {
            lock (_lock)
            {
                AsyncSendMailStatus status = new AsyncSendMailStatus();
                status.TaskId = taskId;
                status.HasError = true;
                status.Status = error;
                status.Completed = true;
                if (!_taskStatus.ContainsKey(taskId))
                {
                    _taskStatus.Add(taskId, status);
                }
            }
        }

        public static AsyncSendMailStatus QueryStatus(string taskId)
        {
            AsyncSendMailStatus task = null;
            lock (_lock)
            {
                if (_taskStatus.ContainsKey(taskId))
                {
                    task = _taskStatus[taskId];
                    if (task.Completed)
                    {
                        _taskStatus.Remove(taskId);
                    }
                }
            }
            return task;
        }

        private static void _sendEmail(AsyncSendMailStatus status, MailTask mailTask)
        {
            try
            {
                var server = mailTask.BuildServer();
                var mail = mailTask.BuildMail();
                var smtp = new SmtpClient();

                // Add event handlers to current SmtpClient instance.
                smtp.OnAuthorized += new SmtpClient.OnAuthorizedEventHandler(OnAuthorized);
                smtp.OnConnected += new SmtpClient.OnConnectedEventHandler(OnConnected);
                smtp.OnSecuring += new SmtpClient.OnSecuringEventHandler(OnSecuring);
                smtp.OnSendingDataStream +=
                    new SmtpClient.OnSendingDataStreamEventHandler(OnSendingDataStream);

                status.Status = "Connecting server ...";
                smtp.Tag = status;
                smtp.SendMail(server, mail);

                status.Status = "Completed";
                status.HasError = false;
                status.Progress = 100;
            }
            catch (Exception ep)
            {
                status.Status = ep.Message;
                status.HasError = true;
                status.Progress = 100;
            }

            status.Completed = true;
        }

        public static void OnSecuring(
            object sender,
            ref bool cancel
        )
        {
            var status = (AsyncSendMailStatus)((SmtpClient)sender).Tag;
            status.Status = "Securing ...";
        }

        public static void OnAuthorized(
            object sender,
            ref bool cancel
        )
        {
            var status = (AsyncSendMailStatus)((SmtpClient)sender).Tag;
            status.Status = "Authorized";
        }


        public static void OnConnected(
            object sender,
            ref bool cancel
        )
        {
            var status = (AsyncSendMailStatus)((SmtpClient)sender).Tag;
            status.Status = "Connected";
        }

        public static void OnSendingDataStream(
            object sender,
            int sent,
            int total,
            ref bool cancel
        )
        {
            var status = (AsyncSendMailStatus)((SmtpClient)sender).Tag;
            if (total == 0)
            {
                return;
            }

            int progress = (sent * 100) / total;
            status.Progress = progress;
            status.Status = string.Format("Sending {0}% ...", progress);
        }
    }
}